# IMPORT DATA, PACKAGES, AND FUNCTIONS

## import the relevant functions and packages

In [1]:
import pandas as pd
import numpy as np

## download the OFAC list from the web

In [2]:
ofac_list_download = pd.read_csv('https://www.treasury.gov/ofac/downloads/sdn.csv', header=None)

In [3]:
ofac_list = ofac_list_download[[0,1,2]]
ofac_list.columns = ['uid', 'name', 'entity_type']
ofac_list = ofac_list.dropna()
ofac_list

uid                       name entity_type
0         36     AEROCARIBBEAN AIRLINES        -0- 
1        173  ANGLO-CARIBBEAN CO., LTD.        -0- 
2        306     BANCO NACIONAL DE CUBA        -0- 
3        424         BOUTIQUE LA MAISON        -0- 
4        475               CASA DE CUBA        -0- 
...      ...                        ...         ...
11029  39709             BUSHEHR PRISON        -0- 
11030  39710       OSTAD, Mohammad Reza  individual
11031  39711  MIRHEYDARY, Mohammad Reza  individual
11032  39783            LATORTUE, Youri  individual
11033  39796               TORNADO CASH        -0- 

[11034 rows x 3 columns]

In [4]:
final_test_cases = pd.read_csv('final_test_cases.csv', encoding = 'gbk')
final_test_cases

Unnamed: 0      UID             Theme                 Category  \
0              0    UID-3  Positive Control              Exact Match   
1              1    UID-3  Positive Control              Exact Match   
2              2    UID-3  Positive Control              Exact Match   
3              3    UID-3  Positive Control              Exact Match   
4              4    UID-3  Positive Control              Exact Match   
...          ...      ...               ...                      ...   
4622         814  UID-541    Name Additions  Name Part in the Middle   
4623         815  UID-541    Name Additions  Name Part in the Middle   
4624         816  UID-541    Name Additions  Name Part in the Middle   
4625         817  UID-541    Name Additions  Name Part in the Middle   
4626         818  UID-541    Name Additions  Name Part in the Middle   

            Sub-category Entity-Type    Test Case ID OFAC List UID  \
0       100% true match       Entity     UID-3 - 217          6936   
1       100% true match       Entity    UID-3 - 6782         26214   
2       100% true match       Entity    UID-3 - 5237         22284   
3       100% true match       Entity    UID-3 - 2858         15188   
4       100% true match       Entity    UID-3 - 7239         27233   
...                  ...         ...             ...           ...   
4622  2 name parts added      Vessel  UID-541 - 4411         19589   
4623  2 name parts added      Vessel  UID-541 - 5475         23162   
4624  2 name parts added      Vessel  UID-541 - 5476         23163   
4625  2 name parts added      Vessel  UID-541 - 5787         23733   
4626  2 name parts added      Vessel  UID-541 - 6250         25314   

                   Original Name                     Test Case Name  
0     JAM'YAH TA'AWUN AL-ISLAMIA         JAM'YAH TA'AWUN AL-ISLAMIA  
1     GLOBOVISION TELE CA, CORP.         GLOBOVISION TELE CA, CORP.  
2           KOREAN PEOPLE'S ARMY               KOREAN PEOPLE'S ARMY  
3            RANCHO LA HERRADURA                RANCHO LA HERRADURA  
4            TECHNO ENERGY, S.A.                TECHNO ENERGY, S.A.  
...                          ...                                ...  
4622                 CHONG RIM 2       CHONG RIM priority cushion 2  
4623                  KUM SONG 7  KUM reservists transmittal SONG 7  
4624                KUM UN SAN 3        KUM seesaw UN SAN buckles 3  
4625                  SAM JONG 1       SAM energizer comment JONG 1  
4626                SHAHR E KORD          SHAHR mills master E KORD  

[4627 rows x 10 columns]

In [5]:
final_test_cases = final_test_cases[['Original Name', 'Test Case Name']]
final_test_cases['Test Type'] = 'Fuzzy Match'
final_test_cases

C:\Users\mdk\AppData\Local\Temp\ipykernel_25876\4085377496.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  final_test_cases['Test Type'] = 'Fuzzy Match'


Original Name                     Test Case Name  \
0     JAM'YAH TA'AWUN AL-ISLAMIA         JAM'YAH TA'AWUN AL-ISLAMIA   
1     GLOBOVISION TELE CA, CORP.         GLOBOVISION TELE CA, CORP.   
2           KOREAN PEOPLE'S ARMY               KOREAN PEOPLE'S ARMY   
3            RANCHO LA HERRADURA                RANCHO LA HERRADURA   
4            TECHNO ENERGY, S.A.                TECHNO ENERGY, S.A.   
...                          ...                                ...   
4622                 CHONG RIM 2       CHONG RIM priority cushion 2   
4623                  KUM SONG 7  KUM reservists transmittal SONG 7   
4624                KUM UN SAN 3        KUM seesaw UN SAN buckles 3   
4625                  SAM JONG 1       SAM energizer comment JONG 1   
4626                SHAHR E KORD          SHAHR mills master E KORD   

        Test Type  
0     Fuzzy Match  
1     Fuzzy Match  
2     Fuzzy Match  
3     Fuzzy Match  
4     Fuzzy Match  
...           ...  
4622  Fuzzy Match  
4623  Fuzzy Match  
4624  Fuzzy Match  
4625  Fuzzy Match  
4626  Fuzzy Match  

[4627 rows x 3 columns]

# FILTER FOR THE REQUIREMENTS OF THE TEST CASE TYPE

## randomly choose 4627 rows

In [6]:
ofac_list_sampled = ofac_list.sample(n = final_test_cases.shape[0]).reset_index(drop = True)
ofac_list_sampled

uid                                 name entity_type
0     30702                            KGB ALPHA        -0- 
1     21914                       ISMA'IL, Ayham  individual
2     19941  MUNSA INTERNATIONAL INVESMENTS S.A.        -0- 
3     24750                 EKT SMART TECHNOLOGY        -0- 
4     19188     INMOBILIARIA MULTI-TIENDAS, S.A.        -0- 
...     ...                                  ...         ...
4622  11601                      ALTUN, Ali Riza  individual
4623  30552                  HASNAIN, Syed Johar  individual
4624  33997                    FARMA IZVORI B.I.        -0- 
4625  10593                    LOAN, Waseem Rauf  individual
4626  33993     MARKOM METAL COMMERCE DOO ZVECAN        -0- 

[4627 rows x 3 columns]

# CREATE THE EXACT MATCH TEST CASES

## create blank test cases table

In [7]:
exact_match_test_cases = pd.DataFrame(columns=['Original Name','Test Case Name', 'Test Type'])
exact_match_test_cases

Empty DataFrame
Columns: [Original Name, Test Case Name, Test Type]
Index: []

In [8]:
exact_match_test_cases['Original Name'] = ofac_list_sampled['name']
exact_match_test_cases['Test Case Name'] = ofac_list_sampled['name']
exact_match_test_cases['Test Type'] = 'Exact Match'
exact_match_test_cases

Original Name  \
0                               KGB ALPHA   
1                          ISMA'IL, Ayham   
2     MUNSA INTERNATIONAL INVESMENTS S.A.   
3                    EKT SMART TECHNOLOGY   
4        INMOBILIARIA MULTI-TIENDAS, S.A.   
...                                   ...   
4622                      ALTUN, Ali Riza   
4623                  HASNAIN, Syed Johar   
4624                    FARMA IZVORI B.I.   
4625                    LOAN, Waseem Rauf   
4626     MARKOM METAL COMMERCE DOO ZVECAN   

                           Test Case Name    Test Type  
0                               KGB ALPHA  Exact Match  
1                          ISMA'IL, Ayham  Exact Match  
2     MUNSA INTERNATIONAL INVESMENTS S.A.  Exact Match  
3                    EKT SMART TECHNOLOGY  Exact Match  
4        INMOBILIARIA MULTI-TIENDAS, S.A.  Exact Match  
...                                   ...          ...  
4622                      ALTUN, Ali Riza  Exact Match  
4623                  HASNAIN, Syed Johar  Exact Match  
4624                    FARMA IZVORI B.I.  Exact Match  
4625                    LOAN, Waseem Rauf  Exact Match  
4626     MARKOM METAL COMMERCE DOO ZVECAN  Exact Match  

[4627 rows x 3 columns]

## concat match test cases

In [9]:
final_test_cases = pd.concat([final_test_cases, exact_match_test_cases])
final_test_cases['Label'] = 'Match'
final_test_cases

Original Name                    Test Case Name  \
0           JAM'YAH TA'AWUN AL-ISLAMIA        JAM'YAH TA'AWUN AL-ISLAMIA   
1           GLOBOVISION TELE CA, CORP.        GLOBOVISION TELE CA, CORP.   
2                 KOREAN PEOPLE'S ARMY              KOREAN PEOPLE'S ARMY   
3                  RANCHO LA HERRADURA               RANCHO LA HERRADURA   
4                  TECHNO ENERGY, S.A.               TECHNO ENERGY, S.A.   
...                                ...                               ...   
4622                   ALTUN, Ali Riza                   ALTUN, Ali Riza   
4623               HASNAIN, Syed Johar               HASNAIN, Syed Johar   
4624                 FARMA IZVORI B.I.                 FARMA IZVORI B.I.   
4625                 LOAN, Waseem Rauf                 LOAN, Waseem Rauf   
4626  MARKOM METAL COMMERCE DOO ZVECAN  MARKOM METAL COMMERCE DOO ZVECAN   

        Test Type  Label  
0     Fuzzy Match  Match  
1     Fuzzy Match  Match  
2     Fuzzy Match  Match  
3     Fuzzy Match  Match  
4     Fuzzy Match  Match  
...           ...    ...  
4622  Exact Match  Match  
4623  Exact Match  Match  
4624  Exact Match  Match  
4625  Exact Match  Match  
4626  Exact Match  Match  

[9254 rows x 4 columns]

# CREATE THE NON-MATCH TEST CASES

## create blank test cases table

In [10]:
non_match_test_cases = pd.DataFrame(columns=['Original Name','Test Case Name'])
non_match_test_cases

Empty DataFrame
Columns: [Original Name, Test Case Name]
Index: []

## run loop to generate the test cases

In [11]:
ofac_list_sampled = ofac_list.sample(n = final_test_cases.shape[0]).reset_index(drop = True)

for index, row in ofac_list_sampled.iterrows():
    original_name = row['name']

    final_test_name = np.random.choice(ofac_list_sampled['name'])
    while (final_test_name == original_name):
        final_test_name = np.random.choice(ofac_list_sampled['name'])
    
    non_match_test_cases.loc[len(non_match_test_cases)] = [row['name'], final_test_name] # append to the dataframe

non_match_test_cases['Test Type'] = 'Non-Match'
non_match_test_cases['Label'] = 'Not Match'
non_match_test_cases

Original Name  \
0                                  VALDEZ BENITES, Joel   
1                                             CARLOTA C   
2         RAMENSKOYE DESIGN COMPANY JOINT STOCK COMPANY   
3                                          NAJIB, Ahmad   
4                           GLEBOVA, Lyubov Nikolayevna   
...                                                 ...   
9249                                 GLOBAL AGE LIMITED   
9250  OBSHCHESTVO S OGRANICHENNOI OTVETSTVENNOSTYU M...   
9251                         AHMED, Qassim Abdullah Ali   
9252                      JINGHO TECHNOLOGY CO. LIMITED   
9253                NHEMA, Chenayaimoyo Dunstan Francis   

                                         Test Case Name  Test Type      Label  
0                                       AL-ASSAD, Karam  Non-Match  Not Match  
1                          SCUADRA FORTIA, S.A. DE C.V.  Non-Match  Not Match  
2                                    AMD CO. LTD AGENCY  Non-Match  Not Match  
3                              WAKED HATUM, Nidal Ahmed  Non-Match  Not Match  
4                                    AKHAEI, Shaghayegh  Non-Match  Not Match  
...                                                 ...        ...        ...  
9249                                 SHAABAN, Bouthaina  Non-Match  Not Match  
9250                     TAMIN KALAYE SABZ ARAS COMPANY  Non-Match  Not Match  
9251               KOREA HAEGUMGANG TRADING CORPORATION  Non-Match  Not Match  
9252  COMITE' DE BIENFAISANCE ET DE SECOURS AUX PALE...  Non-Match  Not Match  
9253                                     BHARDWAJ, Anuj  Non-Match  Not Match  

[9254 rows x 4 columns]

# CONCAT FINAL TEST CASES

In [12]:
final_test_cases = pd.concat([final_test_cases, non_match_test_cases])
final_test_cases

Original Name  \
0                            JAM'YAH TA'AWUN AL-ISLAMIA   
1                            GLOBOVISION TELE CA, CORP.   
2                                  KOREAN PEOPLE'S ARMY   
3                                   RANCHO LA HERRADURA   
4                                   TECHNO ENERGY, S.A.   
...                                                 ...   
9249                                 GLOBAL AGE LIMITED   
9250  OBSHCHESTVO S OGRANICHENNOI OTVETSTVENNOSTYU M...   
9251                         AHMED, Qassim Abdullah Ali   
9252                      JINGHO TECHNOLOGY CO. LIMITED   
9253                NHEMA, Chenayaimoyo Dunstan Francis   

                                         Test Case Name    Test Type  \
0                            JAM'YAH TA'AWUN AL-ISLAMIA  Fuzzy Match   
1                            GLOBOVISION TELE CA, CORP.  Fuzzy Match   
2                                  KOREAN PEOPLE'S ARMY  Fuzzy Match   
3                                   RANCHO LA HERRADURA  Fuzzy Match   
4                                   TECHNO ENERGY, S.A.  Fuzzy Match   
...                                                 ...          ...   
9249                                 SHAABAN, Bouthaina    Non-Match   
9250                     TAMIN KALAYE SABZ ARAS COMPANY    Non-Match   
9251               KOREA HAEGUMGANG TRADING CORPORATION    Non-Match   
9252  COMITE' DE BIENFAISANCE ET DE SECOURS AUX PALE...    Non-Match   
9253                                     BHARDWAJ, Anuj    Non-Match   

          Label  
0         Match  
1         Match  
2         Match  
3         Match  
4         Match  
...         ...  
9249  Not Match  
9250  Not Match  
9251  Not Match  
9252  Not Match  
9253  Not Match  

[18508 rows x 4 columns]

In [13]:
final_test_cases['Label'].value_counts()

Match        9254
Not Match    9254
Name: Label, dtype: int64

In [14]:
final_test_cases['Test Type'].value_counts()

Non-Match      9254
Fuzzy Match    4627
Exact Match    4627
Name: Test Type, dtype: int64

In [15]:
final_test_cases.to_csv('ml_dataset.csv', index = False)